File: services/multiplication-service/internal/infrastructure/in_memory_calculation_cache.go
package infrastructure

import (
	"fmt"
	"multiplication-service/internal/domain"
	"sync"
)

type InMemoryCalculationCache struct {
	cache map[string]*domain.CalculationResult
	mutex sync.RWMutex
}

func NewInMemoryCalculationCache() *InMemoryCalculationCache {
	return &InMemoryCalculationCache{
		cache: make(map[string]*domain.CalculationResult),
	}
}

func (c *InMemoryCalculationCache) Get(request *domain.CalculationRequest) *domain.CalculationResult {
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	
	key := c.generateKey(request)
	return c.cache[key]
}

func (c *InMemoryCalculationCache) Store(request *domain.CalculationRequest, result *domain.CalculationResult) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	
	key := c.generateKey(request)
	c.cache[key] = result
}

func (c *InMemoryCalculationCache) generateKey(request *domain.CalculationRequest) string {
	return fmt.Sprintf("multiply:%s:%s", request.Operand1.String(), request.Operand2.String())
}

File: services/division-service/src/infrastructure/in_memory_calculation_cache.rs
use crate::domain::{CalculationCache, CalculationRequest, CalculationResult};
use std::collections::HashMap;
use std::sync::Mutex;

pub struct InMemoryCalculationCache {
    cache: Mutex<HashMap<String, CalculationResult>>,
}

impl InMemoryCalculationCache {
    pub fn new() -> Self {
        Self {
            cache: Mutex::new(HashMap::new()),
        }
    }

    fn generate_key(&self, request: &CalculationRequest) -> String {
        format!("divide:{}:{}", request.operand1, request.operand2)
    }
}

impl CalculationCache for InMemoryCalculationCache {
    fn get(&self, request: &CalculationRequest) -> Option<CalculationResult> {
        let key = self.generate_key(request);
        let cache = self.cache.lock().unwrap();
        cache.get(&key).cloned()
    }

    fn store(&self, request: &CalculationRequest, result: &CalculationResult) {
        let key = self.generate_key(request);
        let mut cache = self.cache.lock().unwrap();
        cache.insert(key, result.clone());
    }
}